File: src/main/java/org/activecraft/activecraftcore/listener/JoinQuitListener.java
package org.activecraft.activecraftcore.listener;

import org.activecraft.activecraftcore.ActiveCraftCore;
import org.activecraft.activecraftcore.manager.VanishManager;
import org.activecraft.activecraftcore.messagesv2.MessageFormatter;
import org.activecraft.activecraftcore.messagesv2.MessageSupplier;
import org.activecraft.activecraftcore.playermanagement.Playerlist;
import org.activecraft.activecraftcore.playermanagement.Profilev2;
import org.activecraft.activecraftcore.playermanagement.tables.ProfilesTable;
import org.activecraft.activecraftcore.utils.config.MainConfig;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerJoinEvent;
import org.bukkit.event.player.PlayerQuitEvent;
import org.bukkit.event.player.PlayerTeleportEvent;

import java.time.LocalDateTime;

public class JoinQuitListener implements Listener {

    MainConfig mainConfig = ActiveCraftCore.getInstance().getMainConfig();

    @EventHandler(priority = EventPriority.LOWEST)
    public void onPlayerWorldChange(PlayerTeleportEvent event) {
        Player player = event.getPlayer();
        Location playerLocation = player.getLocation();
        Profilev2 profile = Profilev2.of(player);

        profile.getLocationManager().setLastLocation(event.getFrom().getWorld(), playerLocation);
    }

    @EventHandler(priority = EventPriority.LOWEST)
    public void onPlayerJoin(PlayerJoinEvent event) {
        Player player = event.getPlayer();

        Playerlist playerlist = ActiveCraftCore.getInstance().getPlayerlist();

        playerlist.addPlayerIfAbsent(player);

        Profilev2 profile = Profilev2.createIfNotExists(player);
        MessageSupplier acCoreMessageSupplier = profile.getMessageSupplier(ActiveCraftCore.getInstance());

        profile.setTimesJoined(profile.getTimesJoined() + 1);

        playerlist.updateIfChanged(player);

        if (profile.getLocationManager().getLastLocationBeforeQuit() != null)
            player.teleport(profile.getLocationManager().getLastLocationBeforeQuit());

        profile.getDisplayManager().updateDisplayname();
        profile.getEffectManager().updateEffects();

        event.setJoinMessage(null);
        String joinFormat = acCoreMessageSupplier.getFormatted("general.join-format",
                new MessageFormatter(acCoreMessageSupplier.getActiveCraftMessage())
                        .addFormatterPattern("displayname", player.getDisplayName())
                        .addFormatterPattern("playername", player.getName())
        );
        boolean sendJoinMessage = mainConfig.getSendJoinMessage().getValue();

        // vanish stuff
        if (profile.isVanished()) {
            VanishManager.setVanished(profile, true, true);
            if (sendJoinMessage)
                Bukkit.broadcast(joinFormat, "activecraft.vanish.see");
        } else if (sendJoinMessage) {
            event.setJoinMessage(joinFormat);
        }
        if (!player.hasPermission("activecraft.vanish.see")) VanishManager.hideAll(player);

        //fly
        if (profile.isFly()) player.setAllowFlight(true);
    }

    @EventHandler(priority = EventPriority.LOWEST)
    public void onPlayerQuit(PlayerQuitEvent event) {
        Player player = event.getPlayer();
        Location playerLocation = player.getLocation();

        Profilev2 profile = Profilev2.of(player);
        MessageSupplier acCoreMessageSupplier = profile.getMessageSupplier(ActiveCraftCore.getInstance());

        profile.setLastOnline(LocalDateTime.now());
        profile.getLocationManager().setLastLocation(playerLocation.getWorld(), playerLocation, true);
        profile.setBypassLockdown(player.hasPermission("lockdown.bypass"));
        // TODO: 27.08.2022 save profile
        long before = System.nanoTime();
        ProfilesTable.INSTANCE.writeToDatabase(profile);
        long after = System.nanoTime();

        System.out.println("Microseconds: " + ((after - before) / 1000));

        event.setQuitMessage(null);
        String quitFormat = acCoreMessageSupplier.getFormatted("general.quit-format",
                new MessageFormatter(acCoreMessageSupplier.getActiveCraftMessage())
                        .addFormatterPattern("displayname", player.getDisplayName())
                        .addFormatterPattern("playername", player.getName())
        );
        boolean sendQuitMsg = mainConfig.getSendQuitMessage().getValue();

        if (profile.isVanished()) {
            VanishManager.setVanished(profile, false, true);
            if (sendQuitMsg)
                Bukkit.broadcast(quitFormat, "activecraft.vanish.see");
        } else if (sendQuitMsg) {
            event.setQuitMessage(quitFormat.replace("%displayname%", profile.getNickname()));
        }
    }
}

File: src/main/java/org/activecraft/activecraftcore/listener/JoinQuitListener.kt
package org.activecraft.activecraftcore.listener

import org.activecraft.activecraftcore.ActiveCraftCore
import org.activecraft.activecraftcore.manager.VanishManager.hideAll
import org.activecraft.activecraftcore.manager.VanishManager.setVanished
import org.activecraft.activecraftcore.messages.MessageFormatter
import org.activecraft.activecraftcore.messages.MessageSupplier
import org.activecraft.activecraftcore.playermanagement.Playerlist
import org.activecraft.activecraftcore.playermanagement.Profile.Companion.createIfNotExists
import org.activecraft.activecraftcore.playermanagement.Profile.Companion.of
import org.activecraft.activecraftcore.playermanagement.tables.ProfilesTable.writeToDatabase
import org.activecraft.activecraftcore.utils.config.MainConfig
import org.bukkit.Bukkit
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerJoinEvent
import org.bukkit.event.player.PlayerQuitEvent
import org.bukkit.event.player.PlayerTeleportEvent
import java.time.LocalDateTime

class JoinQuitListener : Listener {
    var mainConfig: MainConfig = ActiveCraftCore.INSTANCE.mainConfig
    @EventHandler(priority = EventPriority.LOWEST)
    fun onPlayerWorldChange(event: PlayerTeleportEvent) {
        val player = event.player
        val playerLocation = player.location
        val profile = of(player)
        profile.locationManager.setLastLocation(event.from.world, playerLocation)
    }

    @EventHandler(priority = EventPriority.LOWEST)
    fun onPlayerJoin(event: PlayerJoinEvent) {
        val player = event.player
        val playerlist: Playerlist = ActiveCraftCore.INSTANCE.playerlist
        playerlist.addPlayerIfAbsent(player)
        val profile = createIfNotExists(player)
        val acCoreMessageSupplier: MessageSupplier = profile.getMessageSupplier(ActiveCraftCore.INSTANCE)!!
        profile.timesJoined = profile.timesJoined + 1
        playerlist.updateIfChanged(player)
        if (profile.locationManager.lastLocationBeforeQuit != null) player.teleport(profile.locationManager.lastLocationBeforeQuit!!)
        profile.displayManager.updateDisplayname()
        profile.effectManager.updateEffects()
        event.joinMessage = null
        val joinFormat = acCoreMessageSupplier.getFormatted(
            "general.join-format",
            MessageFormatter(acCoreMessageSupplier.activeCraftMessage)
                .addFormatterPattern("displayname", player.displayName)
                .addFormatterPattern("playername", player.name)
        )
        val sendJoinMessage: Boolean = mainConfig.sendJoinMessage

        // vanish stuff
        if (profile.isVanished) {
            setVanished(profile = profile, hide = true, silent = true)
            if (sendJoinMessage) Bukkit.broadcast(joinFormat, "activecraft.vanish.see")
        } else if (sendJoinMessage) {
            event.joinMessage = joinFormat
        }
        if (!player.hasPermission("activecraft.vanish.see")) hideAll(player)

        //fly
        if (profile.isFly) player.allowFlight = true
    }

    @EventHandler(priority = EventPriority.LOWEST)
    fun onPlayerQuit(event: PlayerQuitEvent) {
        val player = event.player
        val playerLocation = player.location
        val profile = of(player)
        val acCoreMessageSupplier: MessageSupplier = profile.getMessageSupplier(ActiveCraftCore.INSTANCE)!!
        profile.lastOnline = LocalDateTime.now()
        profile.locationManager.setLastLocation(playerLocation.world, playerLocation, true)
        profile.bypassLockdown = player.hasPermission("lockdown.bypass")
        // TODO: 27.08.2022 save profile
        val before = System.nanoTime()
        writeToDatabase(profile)
        val after = System.nanoTime()
        println("Microseconds: " + (after - before) / 1000)
        event.quitMessage = null
        val quitFormat = acCoreMessageSupplier.getFormatted(
            "general.quit-format",
            MessageFormatter(acCoreMessageSupplier.activeCraftMessage)
                .addFormatterPattern("displayname", player.displayName)
                .addFormatterPattern("playername", player.name)
        )
        val sendQuitMsg: Boolean = mainConfig.sendQuitMessage
        if (profile.isVanished) {
            setVanished(profile = profile, hide = false, silent = true)
            if (sendQuitMsg) Bukkit.broadcast(quitFormat, "activecraft.vanish.see")
        } else if (sendQuitMsg) {
            event.quitMessage = quitFormat.replace("%displayname%", profile.nickname)
        }
    }
}